src/bp/core/services/middleware/event-storage.ts

```diff
@@ -0,0 +1,114 @@
+import * as sdk from 'botpress/sdk'
+import { KnexExtension } from 'common/knex'
+import { BotpressConfig } from 'core/config/botpress.config'
+import Database from 'core/database'
+import { EventRepository } from 'core/repositories'
+import { TYPES } from 'core/types'
+import { inject, injectable, tagged } from 'inversify'
+import Knex from 'knex'
+import _ from 'lodash'
+import moment from 'moment'
+import ms from 'ms'
+
+import { SessionIdFactory } from '../dialog/session/id-factory'
+
+@injectable()
+export class EventStorage {
+  private readonly BATCH_SIZE = 10
+  private readonly TABLE_NAME = 'events'
+  private knex!: Knex & KnexExtension
+  private intervalRef
+  private currentPromise
+
+  private enabled = false
+  private interval!: number
+  private retentionPeriod!: number
+  private batch: sdk.IO.StoredEvent[] = []
+  private ignoredTypes: string[] = []
+  private ignoredProperties: string[] = []
+
+  constructor(
+    @inject(TYPES.Logger)
+    @tagged('name', 'EventStorage')
+    private logger: sdk.Logger,
+    @inject(TYPES.EventRepository) private eventRepo: EventRepository
+  ) {}
+
+  async initialize(botpressConfig: BotpressConfig, database: Database) {
+    const config = _.get(botpressConfig, 'eventStorage')
+    if (!config || !config.enabled) {
+      return
+    }
+
+    this.knex = database.knex
+    this.interval = ms(config.storageInterval)
+    this.retentionPeriod = ms(config.retentionPeriod)
+    this.ignoredTypes = config.ignoredEventTypes || []
+    this.ignoredProperties = config.ignoredEventProperties || []
+    this.enabled = true
+  }
+
+  public storeEvent(event: sdk.IO.OutgoingEvent | sdk.IO.IncomingEvent) {
```

Comment:
Let me suggest something a bit different for the event collection.

* Keep the interval thing it's great.
* Remove the batch size and change it as a maximum batch size.
* In the collection job (interval), batch insert the whole batch.
* in the storeEvent function, push the event to the batch and if the batch has reached the maxlenght, collect it.

This would remove the necessity of splice logic.

Reply:
The batch size was definitely too low (I was testing and forgot to reset). I'll set it back to 100. Not sure it would be a good idea, because the storing delay would be wildly variable.

Either a low batch size and we are storing all the time under load, or a big batch size and delays are much longer without load.